Sources/WeaverDI/ThreadSafeInstanceCachingDecorator.swift

```diff
@@ -0,0 +1,27 @@
+//
+//  ThreadSafeInstanceCachingDecorator.swift
+//  WeaverDI
+//
+//  Created by Bruno Mazzo on 5/31/18.
+//
+
+import Foundation
+
+class ThreadSafeInstanceCachingDecorator: InstanceCaching {
+    
+    let cache: InstanceCaching
+    let semaphore: DispatchSemaphore = DispatchSemaphore(value: 1)
```

Comment:
A semaphore blocks all the threads at the same time. This cache is accessed a lot, and I'm afraid a semaphore won't be efficient enough.

How about this?
```swift
    private let queue = DispatchQueue(label: "\(ThreadSafeInstanceCachingDecorator.self)")

    ...

    func cache<T>(for key: InstanceKey, scope: Scope, builder: () -> T) -> T {
        var entry: T?
        
        queue.sync {
            entry = self.cache.cache(for: key, scope: scope, builder: builder)
        }

        assert(entry != nil, "Entry can't theoritically be nil.")
        return entry ?? builder()
    }
```